### PR TITLE
remove pallet::getter from pallet-staking

### PR DESCRIPTION
**Pull request by Zebedeusz**
# Description

Part of https://github.com/paritytech/polkadot-sdk/issues/3326
Removes all pallet::getter occurrences from pallet-staking and replaces them with explicit implementations.
Adds tests to verify that retrieval of affected entities works as expected so via storage::getter.

## Review Notes

1. Traits added to the `derive` attribute are used in tests (either directly or indirectly).
2. The getters had to be placed in a separate impl block since the other one is annotated with `#[pallet::call]` and that requires `#[pallet::call_index(0)]` annotation on each function in that block. So I thought it's better to separate them.

### CODE REVIEW

File: substrate/frame/staking/src/pallet/impls.rs [hunk 1]
```diff
@@ -575,7 +575,7 @@ impl<T: Config> Pallet<T> {
 
 			let era_duration = (now_as_millis_u64.defensive_saturating_sub(active_era_start))
 				.saturated_into::<u64>();
-			let staked = Self::eras_total_stake(&active_era.index);
+			let staked = Self::eras_total_stake(active_era.index);
```
Comment (seadanda) [inline substrate/frame/staking/src/pallet/impls.rs:578]:
```suggestion
			let staked = ErasTotalStake::<T>::get(&active_era.index);
```
Part of the aim here is to replace any usage of `Self::storage()` with `Storage::<T>::get()` in the sdk codebase

Comment (Zebedeusz) [reply substrate/frame/staking/src/pallet/impls.rs]: Now it's done. I replaced these usages everywhere I was able to spot them.

File: substrate/frame/staking/src/pallet/mod.rs [hunk 2]
```diff
@@ -986,6 +967,150 @@ pub mod pallet {
...
+			MinimumValidatorCount::<T>::get()
+		}
+
+		/// Get the validators that may never be slashed or forcibly kicked out.
+		pub fn invulnerables() -> Vec<T::AccountId> {
+			Invulnerables::<T>::get()
+		}
+
+		/// Get the preferences of a given validator.
+		pub fn validators<EncodedAccountId>(account_id: EncodedAccountId) -> ValidatorPrefs
```
Comment (seadanda) [inline substrate/frame/staking/src/pallet/mod.rs:987]: @kianenigma cargo expand shows that FRAME macros actually expand to these (seemingly over-) generic function signatures, is this just to allow args to be passed as a ref or value, or is there another reason? I guess we also need them here to not be a breaking change, or can we make assumptions about usage here?

File: substrate/frame/staking/src/pallet/mod.rs [hunk 3]
```diff
@@ -986,6 +967,156 @@ pub mod pallet {
...
+		}
+
+		/// Get the active era information.
+		pub fn active_era() -> Option<ActiveEraInfo> {
+			ActiveEra::<T>::get()
+		}
+
+		/// Get the session index at which the era starts for the last [`Config::HistoryDepth`]
+		/// eras.
+		pub fn eras_start_session_index<EncodeLikeEra>(
```
Comment (seadanda) [inline substrate/frame/staking/src/pallet/mod.rs:1016]:
```suggestion
		pub fn eras_start_session_index<EncodeLikeEraIndex>(
```

File: substrate/frame/staking/src/pallet/mod.rs [hunk 4]
```diff
@@ -986,6 +967,156 @@ pub mod pallet {
...
+			era_index: EncodeLikeEra,
+		) -> Option<SessionIndex>
+		where
+			EncodeLikeEra: codec::EncodeLike<EraIndex>,
+		{
+			ErasStartSessionIndex::<T>::get(era_index)
+		}
+
+		/// Get the clipped exposure of a given validator at an era.
+		pub fn eras_stakers_clipped<EncodeLikeEra, EncodeLikeAccountId>(
```
Comment (seadanda) [inline substrate/frame/staking/src/pallet/mod.rs:1026]:
```suggestion
		pub fn eras_stakers_clipped<EncodeLikeEraIndex, EncodeLikeAccountId>(
```
And in other places

Comment (Zebedeusz) [reply substrate/frame/staking/src/pallet/mod.rs]: Done

File: substrate/frame/staking/src/pallet/mod.rs (hunk 3 above)
Comment (Zebedeusz) [reply substrate/frame/staking/src/pallet/mod.rs]: Done

**Review by seadanda — APPROVED** [review]

File: substrate/frame/staking/src/pallet/impls.rs [hunk 5]
```diff
@@ -1384,14 +1384,14 @@ impl<T: Config> historical::SessionManager<T::AccountId, Exposure<T::AccountId,
...
-				let current_era = Self::current_era()
+				let current_era = CurrentEra::<T>::get()
 					// Must be some as a new era has been created.
 					.unwrap_or(0);
 
 				validators
 					.into_iter()
 					.map(|v| {
-						let exposure = Self::eras_stakers(current_era, &v);
+						let exposure = ErasStakersClipped::<T>::get(current_era, &v);
```
Comment (gui1117) [inline substrate/frame/staking/src/pallet/impls.rs:1394]: Why is it changed here? It is not a getter.
```suggestion
						let exposure = Self::eras_stakers(current_era, &v);
```

**Review by gui1117 — COMMENTED** [review]
One change but then looks good.

File: substrate/frame/staking/src/pallet/impls.rs (hunk 5 above)
Comment (Zebedeusz) [reply substrate/frame/staking/src/pallet/impls.rs:1394]: I must have thought that eras_stakers resolves to ErasStakersClipped::<T>::get but not looking at it again, it doesn't seem like it. Good catch.

**Pull request merged**